### PR TITLE
[InstCombine] Fold Ext(i1) Pred shr(A, BW - 1) => i1 Pred A s< 0

### LINKED ISSUE #67916: [InstCombine] `!a == (a < 0)` is not optimized
Consider the following functions:
```cpp
int foo(int a)
{
   return !a == (a > 0);
}

int bar(int a)
{
   return (a < 0);
}

int baz(int a)
{
   return !a == (a < 0);
}

int qux(int a)
{
   return (a > 0);
}
```

Both Clang and GCC can optimize `foo` to `bar`. GCC can also optimize `baz` to `qux`, but Clang cannot.
https://godbolt.org/z/exvTE8Gn7
https://alive2.llvm.org/ce/z/LGvASc

### PR DESCRIPTION
**Pull request by XChy**
Resolves #67916 .
This patch folds `Ext(icmp (A, xxx)) Pred shr(A, BW - 1)` into `i1 Pred A s< 0`.
[Alive2](https://alive2.llvm.org/ce/z/k53Xwa).

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -5325,17 +5321,36 @@ Instruction *InstCombinerImpl::foldICmpEquality(ICmpInst &I) {
...
+        if ((ExtOpc == Instruction::ZExt && ShiftOpc == Instruction::LShr) ||
+            (ExtOpc == Instruction::SExt && ShiftOpc == Instruction::AShr)) {
+          Value *Xor = Builder.CreateXor(X, Y, "xor.signbits");
+          Value *R = (Pred == ICmpInst::ICMP_EQ) ? Builder.CreateIsNeg(Xor)
+                                                 : Builder.CreateIsNotNeg(Xor);
+          return replaceInstUsesWith(I, R);
+        }
+      }
+
+      // Transform (X < 0 ==/!= icmp(X)) into (not) xor(X < 0, icmp(X))
```
[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5344]: `(X < 0 ==/!= icmp(X))` -> `(X u>> BitWidth - 1) ==/!= (zext icmp(X)))` seems to be more accurate.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -5325,17 +5321,36 @@ Instruction *InstCombinerImpl::foldICmpEquality(ICmpInst &I) {
...
+      if (match(A, m_c_ICmp(Pred2, m_Value(X), m_Value())) &&
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5345]: `m_Specific(X)`?

Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5345]: This doesn't need to be a compare, it works for any `i1`.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 3]
```diff
@@ -5325,17 +5321,36 @@ Instruction *InstCombinerImpl::foldICmpEquality(ICmpInst &I) {
...
+          ExtOpc == Instruction::ZExt && ShiftOpc == Instruction::LShr) {
+
+        Value *Xor = Builder.CreateXor(Op0, Op1, "xor.ne");
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5348]: You aren't really emitting `xor(X < 0, icmp(X))` here. You are emitting `xor(X u> BitWidth, zext(icmp(X))`
I'd think something like:
`(icmp eq/ne (icmp slt X, 0), icmp(X))` would be simpler tbh.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 2 above)
Comment (XChy) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5345]: Sure, do you also think `not(xor(lshr,zext(i1)))` is better than `lshr == zext(i1)` ?
And whether can we fold `A == B => not(trunc(xor(A,B)))` when the constant ranges of A and B are both {0,1}?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 3 above)
Comment (XChy) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5348]: InstCombineAndOrXor would fold `bitwise(X u> BitWidth, zext(icmp(X))` into `(bitwise (icmp slt X, 0), icmp(X))`.
Here I think we emit directly `xor` would be more general (?) for other rhs besides `icmp`.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5348]: I'd skip the middle-man given that its simpler to emit/think about.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 2 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5345]: I'd just do `x s< 0 eq/ne i1`

**Comment by goldsteinn** [comment]
Second commit message needs to be updated.

**Comment by sftlbcn** [comment]
worth handling other predicates?
https://alive2.llvm.org/ce/z/EJAH6Z

```
define i1 @src(i1 %x, i32 %y) {
  %zx = zext i1 %x to i32
  %cmp1 = lshr i32 %y, 31
  %cmp2 = icmp ule i32 %cmp1, %zx
  ret i1 %cmp2
}
define i1 @tgt(i1 %x, i32 %y) {
  %cmp1 = icmp slt i32 %y, 0
  %cmp2 = icmp ule i1 %cmp1, %x
  ret i1 %cmp2
}
```

**Comment by XChy** [comment]
> worth handling other predicates? https://alive2.llvm.org/ce/z/EJAH6Z

Thanks for reminder, I think that's worthwhile and convenient to implement here.
`zext(i1) pred lshr(A, BW - 1)` -> `i1 pred (A s< 0)`
The pattern will be added later.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 4]
```diff
@@ -5325,17 +5321,37 @@ Instruction *InstCombinerImpl::foldICmpEquality(ICmpInst &I) {
...
+
+      // Transform (X u>> BitWidth - 1 == zext(i1)) into X s< 0 == i1
+      // Transform (X u>> BitWidth - 1 != zext(i1)) into X s< 0 != i1
+      if (A->getType()->getScalarSizeInBits() == 1 && Op0->hasOneUse() &&
+          Op1->hasOneUse() && ExtOpc == Instruction::ZExt &&
+          ShiftOpc == Instruction::LShr) {
+
+        Value *SLTZero =
+            Builder.CreateICmpSLT(X, Constant::getNullValue(X->getType()));
+        SLTZero->takeName(Op0);
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5352]: This doesn't seem necessary.

**Comment by goldsteinn** [comment]
LGTM.

**Comment by nikic** [comment]
Can you please update the patch description and alive proof? It doesn't seem to match what is actually implemented.

**Comment by nikic** [comment]
And I think the tests need an update too? It doesn't look like the most basic case is tested at all (where the zext i1 and lshr are not correlated).

**Comment by XChy** [comment]
New proof: [alive2](https://alive2.llvm.org/ce/z/c7psiD)

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 5]
```diff
@@ -7186,6 +7186,23 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
...
+        (match(Op0, m_LShr(m_Value(X),
+                           m_SpecificIntAllowUndef(
+                               Op0->getType()->getScalarSizeInBits() - 1))) &&
+         Y->getType()->getScalarSizeInBits() == 1)) {
+
+      Value *SLTZero =
+          Builder.CreateICmpSLT(X, Constant::getNullValue(X->getType()));
+      Value *Cmp = Builder.CreateICmp(Pred, SLTZero, Y, I.getName());
+      return replaceInstUsesWith(I, Cmp);
+    }
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7205]: This code was previously located next to https://github.com/llvm/llvm-project/blob/24950fd2f4138caed78fd5fb28decd00474cb6cb/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp#L5385, which seems like a special case of this transform. Would it be possible to extend this one to handle the `SExt + AShr` variant as well, and then drop the other code completely? Or does what we generate here not fold down to the xor pattern that one produces?

Comment (XChy) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7205]: That's correct. [Proof](https://alive2.llvm.org/ce/z/_VBqK7). And this transform just handles special cases which foldXorICmp also handles.

**Comment by XChy** [comment]
Tests about ashr-sext pair will be added later.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**